### ISSUE TITLE
[New build - EXPERIMENTAL] 20w28d: New Rendering Tech.

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#2655]
First batch of hotfixes following the new rendering tech.

Changes:
- The CPU is now doing a minimum amount of work when a brick is added. 
- Performances should be a bit better on the CPU side.

Bugfixes:
- Missing objects when moving camera fast.
- Deletion can trigger a single frame with older brick visible
- In some case deleting a brick delete the wrong brick visually.
#2648 Some NPC behaviour can cause issue.

**Comment by tsunamayo** [comment]
Build is up! Thanks all, tell me if you have any issue, once we get this cleaned I will release in DEFAULT... Cheers

**Comment by alxnns1** [comment]
![image](https://user-images.githubusercontent.com/485571/87054381-f08d5380-c1fa-11ea-8576-e5662b7933d5.png)
![image](https://user-images.githubusercontent.com/485571/87054399-f6833480-c1fa-11ea-9f27-5c137232bb98.png)
looks like its missing the patch notes and version number

**Comment by BrickMacklin** [comment]
Entire chunks are missing from my build whether I have culling on or off. Stickers are also not appearing. This is the same at any distance from the ship. Attached is a picture. 
![image](https://user-images.githubusercontent.com/37836252/87054837-975cfa00-c1d1-11ea-9771-40124420caeb.png)

**Comment by Kaiser-Indrasil** [comment]
Oh man, am I glad that you decided to delegate the new version to the experimental version... there's so many bugs:
- #2636 still persists
- 10FPS when the ship is near the torus
- 50% fps drop (from 100 down to 50) when Occlusion Culling is enabled. That's initially. After letting the game run for a few minutes, it oscillates between 60 and 80 fps.
- the Occlusion Culling-related bug
![Starship_EVO_DYEGEmSTKn](https://user-images.githubusercontent.com/63756057/87056566-b7a2ae00-c1fd-11ea-93fc-1057c203ba8a.png)
There is no hole in the ship's hull but the occlusion tech doesn't render the armour at some degrees.

Also, this is what happens when I first get far from the ship with the Occlusion Culling enabled, then turned off, and then when I got close again:
![Starship_EVO_KEJFcdj7gS](https://user-images.githubusercontent.com/63756057/87056612-c4bf9d00-c1fd-11ea-97be-b6d52e6f13e3.png)
![Starship_EVO_1A3CwRuN5h](https://user-images.githubusercontent.com/63756057/87057073-53341e80-c1fe-11ea-9bcf-ec598feb664c.png)

**Comment by ZachZent** [comment]
The abstract art update

**Comment by tsunamayo** [comment]
@BrickMacklin could you send me your ship so I can investigate? You are seing no decals at all?

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil Occlusion On / Off is not implemented as I cant skip the compute shader stage with the current implementation. Do you have a log file for the performance, there might be some error that could explain the drop. What is the name of your ship again? Thanks

**Comment by tsunamayo** [comment]
so Kaiser your ship is fine, it takes some FPS to display but I guess it was the same before. Occlusion Off is bugged (I havent worked on it), but thats a debug option. When you say 50% fps drop it is compared to what? Before the update?

**Comment by tsunamayo** [comment]
@BrickMacklin could you send me a log? I loaded your ship and it seems fine on my side, with the exception of the greedy unloading at a distance that I still need to fix (that was in previous update too). Decals were fine too... Thanks

**Comment by Kaiser-Indrasil** [comment]
Yeah, 50% drop compared to the main branch of the game 20w28b. Given a few minutes, it oscillates between 60 and 80 fps.
I'n not sure if I've mentioned it yet but the loading time in the beta branch has at least doubled compared to the stable branch.

Oh, and here's the logs:
[output_log.txt](https://github.com/tsunamayo/Starship-EVO/files/4900014/output_log.txt)

Are you able to reproduce the #2636 on your side?

**Comment by tsunamayo** [comment]
I havent looked at 2636

**Comment by tsunamayo** [comment]
where you using occlusion culling in old version?

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil thats a log from the old version. For the loading time I will improve that a bit, right now it is not super optimum, but it still might be a bit slower that the old tech.

**Comment by Kaiser-Indrasil** [comment]
Damn, that is a valid point! With OC enabled, the fps in the Stable Branch oscillate between 80-100.
Oh, so the log file wipes itself each time I start the game?

**Comment by tsunamayo** [comment]
yep it will. For the missing bit, was it with OC off? Occlusion Culling Off was bug.
In your case as your ship is an empty shell and you have a big GPU I dont expect to see any gain with Occlusion Culling, that thats perfectly normal. It is not a tech that will improve your FPS all the time, it is a tech with a fixed cost that garantee your fps wont get smashed in a lot of use case (when there is a lot of overdrawn).

**Comment by BrickMacklin** [comment]
@tsunamayo Here you go! I did some more testing. The decals were appearing with culling turned on but then there's that usual bug of chunks disappearing.
[Paris Frigate.zip](https://github.com/tsunamayo/Starship-EVO/files/4900070/Paris.Frigate.zip)

**Comment by tsunamayo** [comment]
but there are disappearing at a distance right? Also did you tested all at Occlusion Culling On? Thanks

**Comment by BrickMacklin** [comment]
Let me make a video and get back to you

**Comment by Kaiser-Indrasil** [comment]
The one with the red circle I drew over? Nope, it was with OC on.

The other one when I first got away from the ship, then turn the OC off, and then got close again - it's a minor issue since the player won't really toggle the OC on and off constantly during gameplay.

**Comment by tsunamayo** [comment]
I fixed the distance bug btw.

**Comment by BrickMacklin** [comment]
@tsunamayo here is a video I made only a few minutes ago showing the effects on the ship. It is still processing so it might not show right away. 
https://youtu.be/6BwVzwvDtSM

**Comment by tsunamayo** [comment]
Like I said Occlusion Culling Off is not working, dont use it. Distance bug is fixed. But it seems that when you ticked freeze occlusion culling there was none applied.
Could you send me a screenshot of your ship, with F3/Rendering/Show Hi-Z Buffer ?

**Comment by BrickMacklin** [comment]
![image](https://user-images.githubusercontent.com/37836252/87104867-35c77a80-c227-11ea-93b1-bde354046f57.png)

If I move back this happens. This is on the game default loaded settings. 
![image](https://user-images.githubusercontent.com/37836252/87104891-4d9efe80-c227-11ea-88ea-3a1d098ca7b9.png)

And with culling turned off.
![image](https://user-images.githubusercontent.com/37836252/87104909-614a6500-c227-11ea-95e5-c2b822b574db.png)

**Comment by tsunamayo** [comment]
also I would really like to know if some guy with AMD card tested that build! Thanks all!

**Comment by Kaiser-Indrasil** [comment]
ProPeach has one (weird that I can't tag them here...)

**Comment by ProPeach** [comment]
I'll give it a try after lunch! Is there anything specific you want tested, or just how ships look in the new branch?

**Comment by tsunamayo** [comment]
yeah if it works at all, AMD has been giving me a lot of headaches in the past...

**Comment by ProPeach** [comment]
I'm running an RX 580 8GB
Stable Branch OC on - 40ish FPS of a ship heavy scene, although many don't have interiors but there are a lot of bricks on show
![image](https://user-images.githubusercontent.com/27857711/87155163-5f79b380-c2b2-11ea-9dc0-163d47003ad6.png)
With Occlusion culling frozen, you can see not much is actually removed Some of the base on the ground wasn't on the screen so it's not rendered, but none of the far side of the ships in the foreground were un-rendered even though they were not visible.

Lots of ships had chunks missing thanks to to LOD tech though

Experimental OC on - 50-60 FPS stable, all the stickers etc on ships appear fine. I can't see anything wrong with it!
![image](https://user-images.githubusercontent.com/27857711/87155871-a4521a00-c2b3-11ea-8885-399fa3a19bcf.png)
With Occlusion Culling frozen, we can see entities that were out of the players line of sight were culled (notice the ground base and the grey transport with no nose), but again the far side of ships were not culled. Still lots of chunks missing from the LOD, the render distance for this seems really low.

By disabling OC I can get 5-10 FPS more in this particular scene.

Here's the scene if you'd like to test yourself - 
[Save_Data - ProPeach.zip](https://github.com/tsunamayo/Starship-EVO/files/4903175/Save_Data.-.ProPeach.zip)

I'm happy to test anything else that needs it o/

**Comment by tsunamayo** [comment]
Okay great thanks for testing @ProPeach , at least thats that!
